<a href="https://colab.research.google.com/github/ryane-jon/Natural-Language-Processing-Coursework/blob/main/NLP_Coursework.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Natural Language Proccessing (CE4145) Coursework
- Ryan Jones (2208751)
- Dataset Source: [Sentiment Labelled Sentences](https://archive.ics.uci.edu/dataset/331/sentiment+labelled+sentences)

I plan to use various NLP techniques to create a model that can predict the sentiment (Positive or Negative) of various sentences. I also plan to extend this to detect the source of these sentences (Amazon, imdb or yelp)

### Import dataset
(from the download of the dataset in the provided link)

The 3 files are imported (amazon_cells_labelled.txt, imdb_labelled.txt, yelp_labelled.txt)

In [2]:
#used to upload data
import io
from google.colab import files
uploaded = files.upload() #files can be uploaded here, I select the 3 labelled files.

Saving amazon_cells_labelled.txt to amazon_cells_labelled.txt
Saving imdb_labelled.txt to imdb_labelled.txt
Saving yelp_labelled.txt to yelp_labelled.txt


In [34]:
#import stuff to use later
from sklearn.pipeline import Pipeline #let's import the pipeline functionality
from sklearn.feature_extraction.text import CountVectorizer #and we will import a simple pre-processing method
from sklearn.feature_extraction.text import TfidfTransformer #and a representation learner
from sklearn.neighbors import KNeighborsClassifier #and a simple classifier model
from sklearn.model_selection import StratifiedKFold #cross fold is sometimes called k-fold. Calling the stratified version ensures that classes have equal representation across folds
from sklearn.metrics import accuracy_score #import an accuracy metric to tell us how well the model is doing

Split each files lines into strings in arrays to be processed

In [3]:
amazon_data = uploaded['amazon_cells_labelled.txt'].decode('UTF-8').splitlines()
imdb_data = uploaded['imdb_labelled.txt'].decode('UTF-8').splitlines()
yelp_data = uploaded['yelp_labelled.txt'].decode('UTF-8').splitlines()

#View some of the data
for index in range(5):
  print(amazon_data[index])
  print(imdb_data[index])
  print(yelp_data[index])

So there is no way for me to plug it in here in the US unless I go by a converter.	0
A very, very, very slow-moving, aimless movie about a distressed, drifting young man.  	0
Wow... Loved this place.	1
Good case, Excellent value.	1
Not sure who was more lost - the flat characters or the audience, nearly half of whom walked out.  	0
Crust is not good.	0
Great for the jawbone.	1
Attempting artiness with black & white and clever camera angles, the movie disappointed - became even more ridiculous - as the acting was poor and the plot and lines almost non-existent.  	0
Not tasty and the texture was just nasty.	0
Tied to charger for conversations lasting more than 45 minutes.MAJOR PROBLEMS!!	0
Very little music or anything to speak of.  	0
Stopped by during the late May bank holiday off Rick Steve recommendation and loved it.	1
The mic is great.	1
The best scene in the movie was when Gerardo is trying to find a song that keeps running through his head.  	1
The selection on the menu was great

Create a numpy array with 3 collumns: Review, Sentiment, and Source

In [4]:
import numpy as np
reviews = []
sentiments = []
sources = []

for line in amazon_data:
  reviews.append(line[:-1])
  sentiments.append(line[-1])
  sources.append("Amazon")

for line in imdb_data:
  reviews.append(line[:-1])
  sentiments.append(line[-1])
  sources.append("IMDB")

for line in yelp_data:
  reviews.append(line[:-1])
  sentiments.append(line[-1])
  sources.append("Yelp")

data = np.column_stack((reviews, sentiments, sources))
print(data.shape)

(3002, 3)


Test base NLP Pipeline for Sentiment prediction (based on reviews)

In [37]:
x, y, z = data.T  #x=reviews, y=sentiment, z=source
text_clf = Pipeline([ #Pipeline to organise functions
  ('prep', CountVectorizer()), #Count vectorizer
  ('rep', TfidfTransformer()), #representation learning method using tf-idf
  ('mod', KNeighborsClassifier()), #kNN classifier
  ])

acc_score = []

kf = StratifiedKFold(n_splits=5)
for train, test in kf.split(x,y):
  x_train, x_test, y_train, y_test = x[train], x[test], y[train], y[test]
  text_clf.fit(x_train, y_train) #fit to training data
  predictions = text_clf.predict(x_test) #predict on test data
  acc_score.append(accuracy_score(predictions, y_test)) #get accuracy

print("Accuracy:", np.mean(acc_score)) #mean accuracy


/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_split.py:805: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


Accuracy: 0.7661508596783139


Double-checking that data is being represented correctly

In [11]:
for index in range(0,4):
  print(x[index*100])
  print(y[index*100])
  print(z[index*100])

So there is no way for me to plug it in here in the US unless I go by a converter.	
0
Amazon
Integrated seamlessly with the Motorola RAZR phone.	
1
Amazon
The color is even prettier than I thought it would be, and the graphics are incredibly sharp.	
1
Amazon
Sending it back.	
0
Amazon


### Introduce tokenization pre-proccessing into the pipeline

Import packages for tokenization

In [6]:
import nltk #import the natural language toolkit

nltk.download('punkt') #download the package in nltk which supports tokenization
nltk.download('punkt_tab')
nltk.download('stopwords') #download the nltk package for stopwords

from nltk.tokenize import word_tokenize #import the tokenize package
from nltk.corpus import stopwords #import the package from the corpus
from nltk.stem.snowball import SnowballStemmer #import the snowball stemmer (also known as Porter2)

[nltk_data] Downloading package punkt to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.
[nltk_data] Downloading package punkt_tab to /root/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt_tab.zip.
[nltk_data] Downloading package stopwords to /root/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.


Create a tokenize function and add it to the pipeline

In [22]:
from sklearn.base import BaseEstimator, TransformerMixin
class pre_process_tokenize(BaseEstimator, TransformerMixin):

    def __init__(self):
      return None

    def fit(self, X, y=None):
        return self

    def transform(self, X, y=None):
      prep_text = []
      for x in X:
            #basic tokenisation
            token_text = word_tokenize(x)
            #lower casing & punctiuation removal)
            normd_text = [token.lower() for token in token_text]
            #stopword removal
            swr_text = [token for token in normd_text if token not in stopwords.words('english')] #list compression to remove any stopwords from our list
            #stemming
            stemmer = SnowballStemmer("english") #specify we are using the English stemming rules, as other languages are present in toolkit
            prep_text += [[stemmer.stem(word) for word in swr_text]] #list compression for applying the stemmer

      #rejoin the sentences for use of CountVectorizer
      prep_sentences = [" ".join(sentence) for sentence in prep_text]
      return prep_sentences

text_clf = Pipeline([
  ('prep', pre_process_tokenize()),
  ('count', CountVectorizer()),
  ('rep', TfidfTransformer()),
  ('mod', KNeighborsClassifier()),
  ])


acc_score = []

kf = StratifiedKFold(n_splits=5)
for train, test in kf.split(x,y):

  x_train, x_test, y_train, y_test = x[train], x[test], y[train], y[test]

  text_clf.fit(x_train, y_train)
  predictions = text_clf.predict(x_test)
  acc = accuracy_score(predictions, y_test)
  acc_score.append(acc)

print("Accuracy:", np.mean(acc_score))

/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_split.py:805: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


Accuracy: 0.7281913477537438
So there is no way for me to plug it in here in the US unless I go by a converter.	
Good case, Excellent value.	
Great for the jawbone.	
Tied to charger for conversations lasting more than 45 minutes.MAJOR PROBLEMS!!	
The mic is great.	


Tokenisation seems to have reduced the accuracy of the model (which I personally didn't expect).


It's possible that stemming and stop word removal caused some context and nuance to be lost, which was made worse by the fact that the data comes from multiple different sources. E.g something ascociated with a positive sentiment for one source ends up being negative for another.

In [30]:
class pre_process_tokenize2(BaseEstimator, TransformerMixin):

    def __init__(self):
      return None

    def fit(self, X, y=None):
        return self

    def transform(self, X, y=None):
      prep_text = []
      for x in X:
            #basic tokenisation
            token_text = word_tokenize(x)
            #lower casing & punctiuation removal)
            normd_text = [token.lower() for token in token_text]
            #stemming
            stemmer = SnowballStemmer("english") #specify we are using the English stemming rules, as other languages are present in toolkit
            prep_text += [[stemmer.stem(word) for word in normd_text]] #list compression for applying the stemmer

      #rejoin the sentences for use of CountVectorizer
      prep_sentences = [" ".join(sentence) for sentence in prep_text]
      return prep_sentences

text_clf = Pipeline([
  ('prep', pre_process_tokenize2()),
  ('count', CountVectorizer()),
  ('rep', TfidfTransformer()),
  ('mod', KNeighborsClassifier()),
  ])


acc_score = []

kf = StratifiedKFold(n_splits=5)
for train, test in kf.split(x,y):

  x_train, x_test, y_train, y_test = x[train], x[test], y[train], y[test]

  text_clf.fit(x_train, y_train)
  predictions = text_clf.predict(x_test)
  acc = accuracy_score(predictions, y_test)
  acc_score.append(acc)

print("Accuracy:", np.mean(acc_score))

for i in range(5):
  print(x_train[i])

/usr/local/lib/python3.12/dist-packages/sklearn/model_selection/_split.py:805: UserWarning: The least populated class in y has only 1 members, which is less than n_splits=5.
  warnings.warn(


Accuracy: 0.7648175263449806
So there is no way for me to plug it in here in the US unless I go by a converter.	
Good case, Excellent value.	
Great for the jawbone.	
Tied to charger for conversations lasting more than 45 minutes.MAJOR PROBLEMS!!	
The mic is great.	


Testing for different variations of tokenization pre-proccess
- Tokenisation with **lower casing and punctiuation removal** only: 76.0%
- Tokenisation with **lower casing, punctiuation removal** and **stemming**: 76.5%
- Tokenisation with **lower casing, punctiuation removal** and **stopword removal**: 74.5%
- Tokenisation with **stemming** only: 76.5%

Using only the CountVectorizer yielded 76.6%, so I'm abandoning tokenization. (Things like punctiuation and capitalisation might actually be useful for determining sentiment)